### ISSUE TITLE
system, health: "Janitor"

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#14083]
Cockpit could/should have a system "Janitor" that ties in with the health card.

The main idea is to show an persistent overview of upcoming issues before they hit a critical enough level to display in the health card and to provide a means to address these issues.

It could be a new page, linked from the health card, that would show the amount of available storage — among other things — for critical, growing filesystem points, including:

```
/
/var/log/
/var/cache/
/tmp/
/home/
```

(In many systems, these are the same. In some, they're not. Where they're the same, they'd collapse down to /.)

Critically, there should be a way to vacuum logs and set policies to automatically trim the log size.

For manual compacting, it would be the eqivalent of:
```sh
journalctl --vacuum-time=20d
journalctl --vacuum-size=500M
```

For automatic compacting, we could have an interface that adjusts /etc/systemd/journald.conf with flags like [`SystemMaxUse=, SystemKeepFree=, SystemMaxFileSize=, SystemMaxFiles=, RuntimeMaxUse=, RuntimeKeepFree=, RuntimeMaxFileSize=, RuntimeMaxFiles=`](https://freedesktop.org/software/systemd/man/journald.conf.html#SystemMaxUse=)

In addition to logs, a janitor could show the size of and clean various caches, /tmp/, and so on.

It should be arranged in a human logical format instead of a filesystem format, so it could have groups on the page that say: "Log storage", "Cache", "Quota" (for users, including implied quotas) and mechanisms to free up space for each.

**Comment by garrett** [comment]
A lot is kind of storage related, but this is more about the data (and mainly the metadata) than the drives.

**Comment by garrett** [comment]
All of this already exists in system services, in the form of `systemd` and `journalctl`; I'm suggesting that Cockpit provides a friendly interface for these little-known things.

Currently, a lot of people put `/var/` and `/tmp/` on a separate filesystem to make sure their space do not overflow the root `/` partition. They shouldn't _have_ to, unless they want to. But they're trying to make sure their disks aren't completely full, as that will cause a lot of very bad things to happen (such as apps not opening properly, files not being written, logs not being added, and other corruption issues.)

It would be ideal if they could set a policy (and share this policy — so perhaps a bit of Ansible and a shell script exporting is desired here too, much like SELinux) and have the system handle it for them.

It's also good if they could clear these items much like cleaning the cache and cookies in a browser.

**Comment by marusak** [comment]
This is mostly duplicate of #8787 
And I feel the same about it. It would be a cool separate page. That can be preloaded and post statuses.
If we just pick subset of locations to watch for we are going to get request for some additional location, then for configs to set up location and of course config for locations to be ignored :)

Closing in favor of #8787 . I realize this one is trying to be more broad, but what it mentions in details is already mentioned in the other issue.

**Comment by garrett** [comment]
> show an **persistent** overview of upcoming issues before they hit a critical enough level to display in the health card

> available storage — _among other things_ — for

Emphasis on the above quotes.

While freeing up space is a big part of this, it's not everything. For example:

- zombie processes
- a user process hogging resources
- constant network saturation

But the point of differentiation between this and the other is: One is proactive, the other is reactive. That is, this would let you clean up messes. The other would let you fix problems when they arise. There's a bit of a difference, even if they overlap.

This issue proactive; it's for an always-available subpage for performing system clean-up tasks whenever someone might feel like it.

The other issue is reactive; for "oh crap" moments that arise when there's a problem or when there is very nearly a problem.

If we only had the other issue implemented, there wouldn't be a way to get to the journal vacuuming and other tools. We'd need a persistent page for this, which is what _this issue_ is for.

**Comment by garrett** [comment]
I described this in a Jira ticket, speciflcally https://issues.redhat.com/browse/COCKPIT-741.

I posted the first comment, related to showing the disks (especially in error states) @ https://github.com/cockpit-project/cockpit/issues/8787#issuecomment-789855897

My second comment, about cleaning up the disks (related to this issue) is, as follows...

---

It'd be great if we had a journal truncation feature when needed.

If /var/ is getting pretty full:

```
journalctl --disk-usage
```

The journal can be cleaned up too:

```
journalctl
  --vacuum-size=BYTES Reduce disk usage below specified size
  --vacuum-files=INT Leave only the specified number of journal files
  --vacuum-time=TIME Remove journal files older than specified time
```

And /var is where yum/dnf stores package cache:

```
yum clean {metadata,packages,dbcache,expire-cache,all}
```

It would be nice to clean PackageKit as well, but I'm not sure if there's a way to instruct it to clean its cache. There is:

```
sudo pkcon refresh force -c -1
```

...but that only works for enabled repos and it forces a re-fetch of the metadata. (If repos are disabled, it doesn't work. And re-fetching takes more time.)

This might be growing into a "free up space" app. :cry: 

I guess we could have a free up space dialog _instead_ where it abstracts a lot of this away. It could look as simple as this:

---

### Clean cache and journal

☑ Clear package cache
☑ Trim journal
    [ size dropdown ⏷ ][ time dropdown ⏷ ]

(Clean) Cancel

---

(This should be simple enough, no need for a customizable app or page or whatever.)

Clearing the package cache (old metadata, previously downloaded packages, etc.) would run the appropriate command(s).

As for the journal: I'm pretty sure size and time can be used together. And if the size is "unlimited" or time is "complete history" (or somesuch), then it's a no-op for either (or both). If both are a no-op, we should probably have a notice. Perhaps we should default to a generous size and/or time.

If we did have a dialog like this, people may want to active it on their own. So it might need to be a component that's automatically suggested by the overview page's health area and (somehow) able to be manually shown from the Storage page?

**Comment by garrett** [comment]
So what I'm suggesting here is a dialog that would allow someone to clear the cache and trim the journal. It would be available when needed and possibly as an action from somewhere else, such as actions on the storage page somewhere (possibly associated with the system partition which has cache and logs).

In the context of this issue, I'm more concerned about having this reusable component existing for the usage of when #8787 alerts about space not being available. Having it available elsewhere (even when not _needed_) is less of a priority.

**Comment by garrett** [comment]
I tagged this as the Overview page, but it's really a dialog component that will be mainly featured on the overview page (and might need to be called from the Metrics and/or Storage page too).

**Comment by martinpitt** [comment]
I was about to close this as a duplicate of #8787, but that already happened once. I still think this is too broad in scope for a single issue, but kicking it forward another review round then.

**Comment by garrett** [comment]
It's not really too broad; it mainly comes down to two things:

1. A dialog that makes it easy to reclaim space.
2. A link to this dialog shows up in overview when applicable (low disk space health alert) and always in a fixed place (perhaps metrics, in the storage overview and/or in storage, when looking at individual partitions)

The UI to clean things up (# 1) would be in a dialog and contain these elements:

> ### Clean cache and journal
> 
> ☑ Clear package cache
> ☑ Trim journal
>       [ size dropdown ⏷ ][ time dropdown ⏷ ]
> 
> (Clean) Cancel
 
(That's from this comment: https://github.com/cockpit-project/cockpit/issues/14083#issuecomment-789866697)

---

At some point, I did happen to mention zombie processes, processes with high load, and such... but that should probably go elsewhere. Let's keep this focused the critical task of on-disk cleanup.

(I think I mentioned those things as an aside suggesting that the health related tasks are not all about disks, using those as examples of other things we should be covering that are system health related (but not in this PR)? But anyway, this conversation is really about disks running full and needing an easy way to fix that.)

**Comment by garrett** [comment]
And to be _really clear_, this issue is really about # 1: The dialog that cleans up the mess on the disk.

The other issue is about warning when there's a problem. They're linked, but separate.

**Comment by marusak** [comment]
> A dialog that makes it easy to reclaim space.

But that is too broad - I would say it is for a full separate application :) Just offering package cache and journal seems weird and kind of guess work. From my side this is nack, but if more people think that our customers would make use of this (and it won't end up with new and new feature requests to add some different locations like containers, Vms, temp. locations...) then we can do this.

**Comment by martinpitt** [comment]
The low disk space/clean up journal and package cache are exactly what #8787 and https://issues.redhat.com/browse/COCKPIT-741 track (that part is in our quarterly plan). But nevermind, we can revisit these two once that lands.

**Comment by martinpitt** [comment]
I think at least cleaning the package cache is already a great -- and fairly safe -- start. Trimming logs needs to be communicated to the UI fairly well, but as long as it's clear what happens, that seems fine to me as well. It's at least an easy band-aid when your server is about to collapse.

**Comment by garrett** [comment]
It's not guess work. All systems ship with package updating and logging. And these are the two things that continually fill up disks.

I'm suggesting that we can handle these common cases simply, and even cleaning out both type of package caching used on servers: dnf/yum/apt/etc. and packagekit.

Most of the time, the other sources of major disk space usage are:
- containers
- VMs
- databases (nothing one can do really; we cannot — nor should not — delete databases)
- files in /home/ (this is so arbitrary and often not the case for most servers; only for multi-user systems; nothing we can realistically handle)

As for containers and VMs, we should show the space used by each, if it's active or not, and allow someone to delete the containers/images and VMs. But that's way outside the scope of this PR. All of the above list is.

What's inside the scope? Stuff every system has, and stuff every sysadmin has to take note of and battle. And that is: ever-growing logs filling the disk (which is why sometimes sysadmins even put /var/log/ on a separate partition) and caches, mainly from packages. We can make this simple to clean up, and that's what this PR does.

I don't think we need to think about feature creep for this; it's fairly limited to just stuff that grows uncontrollably that's on every system... and what our users told us are a problem. And all of us on the team have probably hit our heads on running out of space from logs and package caches before.

**Comment by garrett** [comment]
> an easy band-aid when your server is about to collapse

Exactly. It's a way to quickly and easily get back significant storage on a system to have a bit of breathing room, to prevent it from going into an emergency mode where the disk is full and stuff starts crashing and corrupting data (including being unable to save things).

**Comment by garrett** [comment]
The reason why this and #8787 are split up are so they can land at different times and be implemented by different people (or the same person).

Sometimes I open a PR that's big and sweeping and needs to be split apart, but I wanted to make sure break this down to two separate tasks, as even just _warning_ about the issue of low disk space is a great first step. Acting upon it (this issue) is a big win too, but a second step really. (Emergency cleanups aren't as necessary when there is enough space.)

**Comment by marcoczen** [comment]
+1
